api/METRICS.md

```diff
@@ -86,21 +86,186 @@ group_max_from_summary(summary)
 
 |metric|variants|task|notes|aif360|
 |------|--------|-----|----|------|
-|`selection_rate`| G,D,R,Min | class | | &#x2713; |
-|`demographic_parity`| D,R | class | `selection_rate_difference`, `selection_rate_ratio` | `statistical_parity_difference`, `disparate_impact`|
-|`accuracy_score`| G,D,R,Min | class | sklearn | `accuracy` |
-|`balanced_accuracy_score` | G | class | sklearn | - |
-|`mean_absolute_error` | G,D,R,Max | class, reg | sklearn | class only: `error_rate` |
 |`confusion_matrix` | G | class | sklearn | `binary_confusion_matrix` |
 |`false_positive_rate` | G,D,R | class | | &#x2713; |
 |`false_negative_rate` | G | class | | &#x2713; |
 |`true_positive_rate` | G,D,R | class | | &#x2713; |
 |`true_negative_rate` | G | class | | &#x2713; |
-|`equalized_odds` | D,R | class | max of difference or ratio under `true_positive_rate`, `false_positive_rate` | - |
-|`precision_score`| G | class | sklearn | &#x2713; |
-|`recall_score`| G | class | sklearn | &#x2713; |
-|`f1_score`| G | class | sklearn | - |
-|`roc_auc_score`| G | prob | sklearn | - |
-|`log_loss`| G | prob | sklearn | - |
-|`mean_squared_error`| G | prob, reg | sklearn | - |
-|`r2_score`| G | reg | sklearn | - |
+|`selection_rate`| G,D,R | class | | &#x2713; |
+|`demographic_parity`| D,R | class | `selection_rate_difference`, `selection_rate_ratio` | `statistical_parity_difference`, `disparate_impact`|
+|`equalized_odds` | D,R | class | max of difference or min ratio under `true_positive_rate`, `false_positive_rate` | - |
+|`accuracy_score`| G,D,R,Min | class | sklearn | `accuracy` |
+|`balanced_accuracy_score` | G,Min | class | sklearn | - |
+|`precision_score`| G,Min | class | sklearn | &#x2713; |
+|`recall_score`| G,Min | class | sklearn | &#x2713; |

```

Comment:
I don't quite understand why some rows have G, D, R, Min, and others don't.
Can't we, for example, calculate the `precision_socre` difference and ratio?

Reply:
I should have explained these better. The table contains the variants that would be automatically provided by `fairlearn.metrics`. The user can easily create all variants via the `make_derived_metric` in the current `fairlearn.metrics`, which composes `<metric>_group_summary` and `<transform>_from_summary`.

In principle, we could create all the variants G,D,R,Min,Max for all the metrics, but I don't think it makes sense. The rationale is as follows:
* for the base utility metrics (selection rate and false/true positive/negative rates) it is sensible to require parity (meaning equality) as a fairness criterion in the context of allocation decisions; therefore difference and ratio are reasonable metrics--they quantify departure from equality
* for the base performance metrics (various score/loss/error metrics), I have a hard time justifying the equality / parity as a fairness criterion, so I'm only suggesting to create the `group_min` or `group_max` variants, corresponding to min-max fairness (i.e., the worst-case across the groups); the exceptions are accuracy rate and error rate (mean absolute error), because I've seen these defined in other packages, but I'm not sold that we need these two exceptions
* there are some additional descriptive metrics (mean prediction) where I'm not sure we need any derived metrics

Thoughts on this? At the minimum, I'll add the explanation and remove the weird exceptions. The question is whether it makes sense to separate out the three kinds of metrics ("performance", "utility", "descriptive") and for each group only provide some of the variants, or would it be better to provide all transformations although some of them are non-sensical.

Reply:
Thanks for elaborating!
I guess it depends on whether you need to create all of them manually, or whether it's a generic way (and users can therefore do whatever they want). You're saying that `make_derived_metric` allows you to do the latter, so I don't think we need to provide more than that, especially for some that you refer to as "non-sensical".

Reply:
yep: see https://github.com/fairlearn/fairlearn/blob/057d86ba174327ab8f2775678400117f20b6a290/fairlearn/metrics/_metrics_engine.py#L191

Reply:
Definitely agree on not creating 'nonsensical' metrics just to have a nice cartesian product.